[4-9단계 - 방탈출 예약 관리] 히스타(오승현) 미션 제출합니다.

Hacanna42 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
https://prolog.techcourse.co.kr/studylogs/4189


## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
1. 계층간 분리가 조금 낯설어서, 이렇게 하는 게 맞는지 그리고 잘 되었는지 궁금합니다.
2. 제가 사용하지 못한 Spring 기능이 있나요? 지금 Spring을 잘 사용하고 있는 걸까요?
3. 지금 수준에서 Service가 필요할까요? Service가 필요할 정도로 Controller가 크다고 느껴지지 않아요. 제가 너무 오버엔지니어링을 경계하는 걸까요?
4. Fake Repository 객체를 만들어서 Controller를 테스트한 방식에 대해 어떻게 생각하시나요?
5. API의 Response 형식이 복잡해지면서 Response 객체를 만들었어요(DTO). 형식을 맞추기 위해 도메인 객체의 구조를 변경하는 것보다는 나은 선택이라고 판단했어요.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Hacanna42 committed:
refactor: List#removeIf 로 코드 가독성 개선

Hacanna42 committed:
test: AssertJ 메서드를 이용해 테스트 코드 리팩터링

Hacanna42 committed:
feat: 기존 로직에 H2 데이터베이스 적용

Hacanna42 committed:
refactor: 영속성 레이어와 컨트롤러 분리

Hacanna42 committed:
feat: 7단계 - 시간 관리 기능 추가

Hacanna42 committed:
test: 테스트 코드 추가

- 미션 테스트
- 페이지 연결 테스트
- ReservationTime 검증 테스트
- Reservation 검증 테스트

Hacanna42 committed:
refactor: Repository와 Controller 계층 분리 개선

Hacanna42 committed:
test: ReservationController 테스트 코드 작성

Hacanna42 committed:
test: ReservationTimeController 테스트 코드 작성

Hacanna42 committed:
feat: 8단계 - 예약과 시간 관리 기능 추가

- 시간 테이블에 저장된 값만 시간으로 사용할 수 있도록 수정
- Reservation 테이블이 직접적으로 시간을 가지지 않고, ReservationTime 테이블을 사용하도록 변경

Hacanna42 committed:
test: 변경된 API에 따라 테스트 코드 수정

Hacanna42 committed:
refactor: 테스트 패키지 구조 변경

Hacanna42 committed:
docs: README.md에 사이트 설명 추가 및 업데이트

sinb57 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -10,32 +10,48 @@
+        return ResponseEntity.ok(ReservationResponse.of(createdReservation, reservationTimeResponse));
+    }
+
+    @GetMapping()
```
```suggestion
    @GetMapping
```
값이 없는 경우 `()`를 생략할 수 있어요~

Hacanna42 replied:
수정하겠습니다. 꼼꼼하게 짚어주셔서 감사합니다 :)

sinb57 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -10,32 +10,48 @@
-    ResponseEntity<Reservation> create(@RequestBody Reservation reservation) {
-        Reservation newReservation = Reservation.toEntity(reservation, index.getAndIncrement());
-        reservations.add(newReservation);
+    ResponseEntity<ReservationResponse> create(@RequestBody Reservation reservation) {
+        Reservation createdReservation = reservationRepository.createReservation(reservation);
+        ReservationTime reservationTime = reservationTimeRepository.readReservationTime(createdReservation.getTimeId());
+
+        ReservationTimeResponse reservationTimeResponse = ReservationTimeResponse.of(reservationTime);
+        return ResponseEntity.ok(ReservationResponse.of(createdReservation, reservationTimeResponse));
+    }
```
> 계층간 분리가 조금 낯설어서, 이렇게 하는 게 맞는지 그리고 잘 되었는지 궁금합니다.

히스타는 레이어드 아키텍처에서 각 계층의 역할과 책임이 무엇이라 생각하시나요?

> 지금 수준에서 Service가 필요할까요? Service가 필요할 정도로 Controller가 크다고 느껴지지 않아요. 제가 너무 오버엔지니어링을 경계하는 걸까요?

로직이 간단하다면 Service를 두지 않고 Controller에서 로직을 수행하기도 해요. 하지만 지금은 학습 단계이기 때문에 Service를 분리하여 장단점을 체감해보셨으면 좋겠어요~ 히스타의 오버엔지니어링 기준은 무엇인가요?

Hacanna42 replied:
> 히스타는 레이어드 아키텍처에서 각 계층의 역할과 책임이 무엇이라 생각하시나요?

기본적으로 레이어드 아키텍처는 주로 변경이 함께 일어나는 책임 단위를 계층으로 분리함으로써, 결합도를 낮추고 응집도를 높일 수 있다는 장점이 있다고 생각합니다.

`Service`로 분리해보는 것도 학습 관점에서 좋은 경험이 되겠네요. 분리해보면서 어떤 점이 좋아지고 무엇이 번거로워지는지 확인해보겠습니다 :)

> 히스타의 오버엔지니어링 기준은 무엇인가요?

현재 시스템이 잘 동작하고, 설계상 문제점도 특별히 부각되지 않으며, 쉽게 예측 가능한 미래의 변경이 보이지 않는데에도 설계를 변경하거나 추가하는 것이 오버엔지니어링이라고 판단하는 것 같습니다.

오버엔지니어링을 경계해야 하는 이유는 오히려 코드를 복잡하게 만들고, 특히 시간이라는 리소스를 사용하기 때문이라고 생각해요. 하지만 현재 상황에서 `Service`를 분리하는 것이 오버엔지니어링이라고 확신하고 있진 않습니다. 관습처럼 사용되는 레이어라서 구조를 복잡하게 만든다고 보진 않기 때문이에요.

단지 현재 기능이 모두 잘 작동되고, 미션의 요구사항이 모두 달성되었으므로 예측되는 요구사항의 변경이 없기 때문에 추가하지 않았어요.
만약 이게 실무 프로젝트였다면 주어진 요구사항을 모두 구현했어도, 이후 복잡한 비즈니스 로직이 추가될 수 있음을 예측하기에 `Service`를 분리할 것 같지만, 지금처럼 분리를 안해놓고 필요할 때 분리하더라도 아무 상관 없다는 생각입니다.

sinb57 commented on src/main/java/roomescape/dto/ReservationResponse.java:
```diff
@@ -0,0 +1,22 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import roomescape.domain.Reservation;
+
+public record ReservationResponse(
```
> API의 Response 형식이 복잡해지면서 Response 객체를 만들었어요(DTO). 형식을 맞추기 위해 도메인 객체의 구조를 변경하는 것보다는 나은 선택이라고 판단했어요.

Request 객체는 분리할 필요가 없을까요? Response 객체를 분리하는 이유는 단순히 필드의 형식을 맞추기 위함인가요?

Hacanna42 replied:
>  Response 객체를 분리하는 이유는 단순히 필드의 형식을 맞추기 위함인가요?

먼저 Response 객체를 분리하게된 주요 계기는, API 응답 형식의 변경이 도메인 객체에 전파되지 않도록 하기 위해서입니다.

첫 구현에서는 Response 객체를 분리하지 않았는데, 8단계를 진행하면서 생긴 요구사항 변경이 도메인 객체까지 전파되는 걸 관찰하고 나서 Response 객체를 추가하게 되었습니다. 따라서 Response 객체는 단순히 필드의 형식을 맞추기 위해 분리했고, 덕분에 응답 관련 변경 사항이 도메인 객체에 전파되는 것을 막았습니다.

>  Request 객체는 분리할 필요가 없을까요?

Request 형식도 충분히 변경될 수 있으므로, Request 객체를 분리하면 입력 형식이 변경 되었을 때 도메인 객체로의 변경 전파를 막을 수 있을 것 같습니다.

즉, Request 객체도 분리할 필요가 있습니다.

sinb57 replied:
> API 응답 형식의 변경이 도메인 객체에 전파되지 않도록 하기 위해서입니다.

Request / Response DTO 분리의 목적을 잘 알고 계시네요 👍

sinb57 commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,3 @@
+spring.h2.console.enabled=true
+spring.h2.console.path=/h2-console
+spring.datasource.url=jdbc:h2:mem:database
```
MySQL이 아닌 h2로 작업하신 이유가 궁금해요~

Hacanna42 replied:
H2 데이터베이스 사용은 미션 요구사항이었습니다.
하지만 선택의 여지가 있었더라도, 동일하게 H2를 선택했을 것 같아요.

H2는 인메모리 DB로 별도의 외부 DB 서버를 구성하지 않아도 되기 때문에, 개발/테스트 단계에서 간편하고 빠른 피드백을 제공해준다는 점이 큰 장점이라고 생각합니다.

sinb57 replied:
인메모리 방식은 어플리케이션 재구동 시 데이터가 휘발되는 단점이 있는데요, H2를 사용함에도 데이터가 휘발되지 않도록 할 수 있어요~ 다른 방식도 찾아보시겠어요?

Hacanna42 replied:
H2의 데이터가 휘발되지 않도록 하는 옵션이 있다는 건 알고 있었는데, Server Mode/In-memory Mode/Embedded Mode 이렇게 3가지의 연결 모드가 있다는 건 처음 알았네요.

docs 참고해서 embedded mode를 사용하도록 설정했습니다!
이랬더니 추가로 테스트 코드에서 휘발성이 아닌 DB에 접근할 수 없어서 오류가 났네요. `application-test.properties` 파일을 작성하고 `@ActiveProfiles("test")` 어노테이션을 테스트 클래스에 붙이니 해결됐습니다. 해당 어노테이션에 대해 추후 학습해보겠습니다 :) 감사합니다 !

sinb57 commented on src/main/java/roomescape/repository/impl/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,66 @@
+import roomescape.repository.ReservationRepository;
+
+@Repository
+public class ReservationRepositoryImpl implements ReservationRepository {
```
`Impl`은 무슨 의미인가요? 만약 또다른 구현체가 프로덕션에 생겨야 한다면, 그 친구의 이름은 어떻게 하실건가요?

Hacanna42 replied:
`Implementation(구현체)` 의 약자입니다.
사실, Fake 객체로 테스트를 하려면 다형성을 활용해야하고, Repostiory를 Interface로 묶어내야 했어요.
막상 묶으니까 실제 프로덕션 구현체의 이름을 짓기가 조금 애매하더라고요. `H2ReservationRepository`, `DBReservationRepository`.. 등의 이름을 고려하다가, 예전에 어디선가 `Impl`키워드를 붙여 사용하는 걸 본 기억이 나서 관성적으로 사용한 것 같아요.

현재 시점에서 해당 네이밍이 크게 부적절하다고 판단되진 않아요. 구현체가 어차피 하나밖에 없기 때문에 `Impl` 이라는 네이밍이 혼동을 줄 여지도 부족하다고 생각합니다.

만약 또 다른 구현체가 생긴다면, 기존 구현체와 새로 생긴 구현체의 가장 큰 차이점을 기준으로 클래스 네임을 작성하지 않을까 생각합니다.

solarsalt replied:
@Hacanna42 대부분의 경우 인터페이스는 구현체 클래스를 갖게되는데요, 그러면 구현체 클래스마다 'Impl' postfix를 붙여주는게 괜찮은 방식인걸까요? 💬 (repository말고도 다른 인터페이스에서까지 확장해서 고민해볼 만한 주제같아요 😆)
다른 크루들과도 얘기해보면 재밌을 것 같아요~

sinb57 replied:
`Impl`은 너무나도 추상적이기 때문에 지양하는 것이 옳다고 봐요. 지금의 경우, 현재 기술에 초점을 맞추어 `Jdbc` 네이밍을 결합하는 것이 좋아보여요

Hacanna42 replied:
다양한 크루들과 해당 주제에 대해 이야기를 많이 나눴는데요. 다양한 크루들과 이프의 피드백 덕분에 저도 설득되었습니다 :)
"현재 구현체만이 가지는 가장 큰 특징으로 네이밍을 하자!" 라고 생각하게 되었고, `JdbcReservationRepository`로 수정했습니다.

sinb57 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.List;
+import roomescape.domain.Reservation;
+
+public interface ReservationRepository {
```
`Repository`는 무엇을 의미하나요?

Hacanna42 replied:
`Repository`는 말 그대로 저장소를 의미하고, MVC 아키텍처에서 바라보는 `Repository`는 공통된 역할이 있습니다.
`Repository`는 영속성 계층에 해당합니다. 여기서 영속성이란 "영구 저장 속성" 즉 프로그램을 종료해도 데이터가 사라지지 않는 속성을 말합니다.

따라서 `Repository`의 역할은 객체를 영구 저장소(DB 등)에 저장하거나, 저장되어 있던 것을 실제 인스턴스로 꺼내오는 역할을 합니다.

sinb57 replied:
지금의 `Repository`는 영속성을 지키고 있나요? `프로그램을 종료해도 데이터가 사라지지 않는 속성`에 따르면, 지금의 H2는 데이터가 휘발되지 않나요? 데이터가 휘발되는 관점에서 보면 일급컬렉션과 무슨 차이가 있는지 궁금한데요. 히스타는 `Repository`와 일급컬렉션의 관계에 대해 어떻게 생각하시는지 궁금해요~

Hacanna42 replied:
> 지금의 Repository는 영속성을 지키고 있나요?

`Repository`는 `Layered Architecture`에 따르면 영속성 계층으로 분류되지만, 현재 H2의 In-memory 모드를 사용하는 `Repository`는 엄밀하게 말하면 영속성을 지키고 있지 않네요. 물론 모드를 변경하는 것으로 쉽게 해결할 수 있어서 이것에 대한 의문점을 제기하진 않았습니다.
<br>

> 히스타는 Repository와 일급컬렉션의 관계에 대해 어떻게 생각하시는지 궁금해요~

`Repository`는 단순히 DB에 접근하는 `DAO(Data Access Object)` 보다는 좀 더 추상적인 상위 개념이라고 이해하고 있습니다.
`Repository`는 말 그대로 저장소라는 뜻이기 때문에, 내부적으론 `DAO` 혹은 `JDBC`를 이용해 DB에 접근하지만, 꼭 DB 테이블의 데이터와 1:1 로 매칭되지 않는 데이터라도 내부적으로 가공 과정을 거쳐 마치 저장소에서 가볍게 꺼내는 듯한 기능을 제공하는 등 `Repository`를 외부에서 사용하는 입장에서 데이터 조작 객체로 보이지 않고 하나의 저장소처럼 보인다는 점에서 일급 컬렉션과 유사한 부분이 있는 것 같아요 :)

sinb57 replied:
정리해주신 덕분에 히스타의 생각 잘 읽었어요~ㅎ 답변을 보면서 확인하고 싶은 부분이 생겨서 질문 남겨보아요~

> 내부적으론 `DAO` 혹은 `JDBC`를 이용해 DB에 접근하지만

`DAO`와 `JDBC`를 동급으로 보는지 궁금해요~

> Repository를 외부에서 사용하는 입장에서 데이터 조작 객체로 보이지 않고 하나의 저장소처럼 보인다는 점에서 일급 컬렉션과 유사한 부분이 있는 것 같아요 :)

그렇다면 `일급 컬렉션`과 `Repository`를 상하 관계로 보시나요? `일급 컬렉션`을 In-memory 방식의 `Repository`로 보시는지 궁금해요~

sinb57 replied:
정리해주신 덕분에 히스타의 생각 잘 읽었어요~ㅎ 답변을 보면서 확인하고 싶은 부분이 생겨서 질문 남겨보아요 ~

> 내부적으론 `DAO` 혹은 `JDBC`를 이용해 DB에 접근하지만

`DAO`와 `JDBC`를 동급으로 보는지 궁금해요~

> Repository를 외부에서 사용하는 입장에서 데이터 조작 객체로 보이지 않고 하나의 저장소처럼 보인다는 점에서 일급 컬렉션과 유사한 부분이 있는 것 같아요 :)

그렇다면 `일급 컬렉션`과 `Repository`를 상하 관계로 보시나요? `일급 컬렉션`을 In-memory 방식의 `Repository`로 보시는지 궁금해요~

Hacanna42 replied:
> DAO와 JDBC를 동급으로 보는지 궁금해요~

`JDBC`는 저수준의 `DB` 접근을 위한 API입니다.
`DAO`는 그것을 활용해서 데이터에 접근하는 객체(패턴) 입니다. `JDBC`는 단순히 접근만 한다면, `DAO`는 비교적 고수준의 행위(예를 들면, 접근 이후 실제 인스턴스로 변환하는 행위 등) 를 합니다. 즉 동급이라고 보진 않습니다.

<br>

> 일급 컬렉션을 In-memory 방식의 Repository로 보시는지 궁금해요~

일급 컬렉션은 컬렉션 형태로 하나의 자료형태에 대해서만 접근할 수 있도록 하기 때문에, Repository랑은 다른 것 같아요. 하지만 외부에서 데이터에 쉽게 접근할 수 있는 저장소처럼 보인다는 점이 비슷한 것 같습니다!

sinb57 commented on src/main/java/roomescape/repository/ReservationTimeRepository.java:
```diff
@@ -0,0 +1,11 @@
+public interface ReservationTimeRepository {
+    ReservationTime createReservationTime(ReservationTime reservationTime);
+    List<ReservationTime> readReservationTimes();
+    ReservationTime readReservationTime(Long id);
```
조회에 실패하여 null이 반환될 수 있는데요. NPE는 가급적 방어해야 해요~ Optional을 활용해보면 어떨까요

Hacanna42 replied:
```java
    public ReservationTime readReservationTime(Long timeId) {
        final String query = "SELECT id, start_at FROM reservation_time WHERE id = ?";
        return jdbcTemplate.queryForObject(query, (resultSet, rowNum) -> new ReservationTime(
                resultSet.getLong("id"),
                resultSet.getTime("start_at").toLocalTime()
        ), timeId);
    }
```

~~위 코드는 `ReservationTimeRepositoryImpl` 클래스의 메서드입니다. 굳이 `Optional`을 사용하지 않고 결과 값을 체크하고 동작하도록 해도 괜찮을 것 같은데 `Optional`을 사용하는 이유가 뭔가요?~~

~~상황에 따라 `null`에 대한 처리를 `Controller`나 `Repository` 수준이 아닌, 비즈니스 로직 수준에서 처리해야 할 일이 있기 때문에, 그리고 인터페이스의 일관성을 지키기 위해 몽땅 `Optional`로 반환하는 걸까요?~~


---

추가적으로, 호출자에게 `null` 처리를 강제하기 위해 `Optional`을 쓰는 것에 대해서도 예전에 고민을 많이 했었는데요. 결론은 `Optional`도 객체이기 때문에 결국에 `null`일 수 있고, 그러면 결국에 중복된 `null` 체크를 강요하는 것 아닐까..? 라고 생각했어요.
이프는 이것에 대해 어떻게 생각하시나요?

Hacanna42 replied:
\+ 조금 더 생각해보니, 단일 객체를 반환하는 `readReservationTime()` 에만 반환 값에 `Optional` 을 사용하면 좋을 것 같네요.

컬렉션을 반환하는 `reads()`는 빈 컬렉션으로 "조회 결과 없음"을 표현할 수 있고, 
`create()` 는 정상적으로 `Request`가 들어왔다면 null이 불가능하지만

`readReservationTime()` 는 정상적인 요청이 들어와도 조회 결과가 없을 수 있고, 조회 결과가 없음이 꼭 프로그램 예외를 나타내는 것이 아니라, 조회 결과가 존재하지 않는 것 자체를 조회 결과로 해석할 수 있겠네요.

따라서 조회 결과가 존재하지 않다는 걸 `null`로 표현하기 보다는, `Optional`로 표현해서 외부 처리를 강제하는 게 좋은 선택이 될 수 있겠네요.

sinb57 replied:
앗! `ReservationTime readReservationTime(Long id);`에 대한 코멘트였어요~ 코드 범위 선택이 잘못되어 있네요,, (분명 한줄만 찝었는데 왜 `ReservationTimeRepository` 인터페이스를 모두 잡는지 모르겠군요,,)

히스타가 답변 주신 `readReservationTime() 는 정상적인 요청이 들어와도 조회 결과가 없을 수 있고, 조회 결과가 없음이 꼭 프로그램 예외를 나타내는 것이 아니라, 조회 결과가 존재하지 않는 것 자체를 조회 결과로 해석할 수 있겠네요.` 내용이 옳아요. Optional을 사용하는 이유는 nullable 하다는 것을 의도적으로 명시하기 위함이에요. 자바는 코틀린과 같은 다른 언어들과 달리 원시타입을 제외한 모든 타입이 nullable 해요. 그렇기 때문에 본래라면 모든 변수 값에 대해 항상 null check 해야 하는데요. 매번 그럴 수 없으니 어느정도 믿고서 사용하게 되지만, nullable 의 가능성이 명확히 보이는 상황에서는 개발 실수가 발생하지 않도록 Optional을 활용하곤 해요~

> Optional도 객체이기 때문에 결국에 null일 수 있고, 그러면 결국에 중복된 null 체크를 강요하는 것 아닐까..? 라고 생각했어요.

맞아요. Optional도 객체이기 때문에 null일 수 있어요. 때문에 Optional에 대한 안티패턴을 찾아보시면 메서드 파라미터로 받지 말라는 내용을 발견하실 수 있는데요. 값을 넘겨받는 입장에서는 Optional 조차도 null일 수 있기 때문에 신뢰할 수 없어요. 하지만 반환 타입이 Optional 인 경우는 null을 감싼 객체이기 때문에 null 이 아니라고 확신할 수 있어요. 물론 개발 실수로 null이 나올 수 있지만요,,

sinb57 commented on src/main/java/roomescape/domain/ReservationTimes.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReservationTimes {
```
테스트 환경에서만 사용되네요! 프로덕션 환경에 남겨둬야 할까요??

Hacanna42 replied:
프로덕션 환경에 남겨둘 필요가 전혀 없습니다.
이전 step에서는 H2 데이터베이스를 적용하지 않아서 잔재가 남게 되었어요.
삭제할까 말까 고민하다가, 언젠간 일급 컬렉션이 필요하지 않을까 ..? 라는 생각으로 남겨뒀는데.. 사실 이런 생각은 위험한 것 같아요 🥺

삭제하겠습니다.

sinb57 commented on src/test/java/roomescape/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+        ResponseEntity<ReservationResponse> actualResponse = reservationController.create(reservation);
+
+        // then
+        Assertions.assertThat(actualResponse).isNotNull();
```
```suggestion
        assertThat(actualResponse).isNotNull();
```

`Assertions.`가 반복되네요. stati import 처리하면 어떨까요?

Hacanna42 replied:
좋은 생각입니다 :)
근데 저는 `Assertions.` 을 반복해서 붙이는 게 귀찮고 어느정도의 노력이 필요하다고 느껴본적이 없어서 사실 굳이 `static import`를 해야할까? 싶긴 해요. 사실 매 테스트 클래스를 생성할때마다 `static import`를 붙이는 게 더 번거롭다고 생각했어요.

만약 컨벤션이라면 따르겠지만 아직까지 큰 불편함을 느끼지 못했고, `static import`보다 클래스명을 앞에 적어주는 방식이 명확성도 높다고 생각하기에 일단은 수정하지 않겠습니다.

sinb57 replied:
`static import`를 하면 IDE가 이를 인식해서 연관된 메서드를 보여주기도 해요. 생산성을 고려해보면 `static import`가 좋아보이는데, 어떻게 생각하시나요?


<img width="684" alt="image" src="https://github.com/user-attachments/assets/bbbe5cb0-f014-49b4-bc78-ba0581f6fc10" />


<img width="689" alt="image" src="https://github.com/user-attachments/assets/24ce301a-5549-4a9b-bc46-53744e36444b" />

Hacanna42 replied:
자동으로 연관된 메서드를 보여주는 기능은 `static import` 없이도 동작하지 않나요?
제가 생각했을 때 `static import`의 좋은 점은 모든 `assertJ`의 코드 앞에 `Assertions.` 을 안 붙여도 되는 게 좋은 점인 것 같아요! (즉 타자수가 적다)

그치만 저는 아직까지는 앞에 `Assertions.` 를 붙이는 게 시간이나 번거로움이 든다고 느낀 적이 없어요.
 한번 다음 미션 때 `static import`를 적용해서 써보고 차이를 느껴볼게요 :) ㅎㅎ

sinb57 replied:
오?? `static import` 없이도 동작하나요?? 안된다는 점을 보이기 위해 위에 스샷을 첨부하긴 했는데, 음 IntelliJ 버전 차이려나요

Hacanna42 replied:
`As` 까지 쳤을 때에는 `Assertions..` `Assert`등이 떴다가.
`Assertions.` 를 해야만 `Assertions`에 관련된 내용들이 뜨긴 해요~

sinb57 replied:
그렇다면 히스타는 매번 `Assertions.`까지 입력한 후 원하는 메서드를 선택하고 있었군요. 개인 취향이기 때문에 스레드는 이만 종료하겠지만, 그러한 비용까지도 절약하면 생산성이 오를 것 같네요~

sinb57 commented on src/main/java/roomescape/repository/impl/ReservationTimeRepositoryImpl.java:
```diff
@@ -0,0 +1,63 @@
+import roomescape.repository.ReservationTimeRepository;
+
+@Repository
+public class ReservationTimeRepositoryImpl implements ReservationTimeRepository {
```
Repository 구현체에 대한 테스트는 불필요할까요?

Hacanna42 replied:
`Repository` 구현체에 대한 테스트도 필요하다고 생각합니다. 실제로 내가 원하는 데이터가 영속적으로 저장되었는지 확인 할 필요가 있으니까요!

하지만, 실제 프로덕션 DB에 접근하면서 테스트를 할 수 없으니, 고민해보다가 이렇게 테스트가 없는 상태로 제출하게 된 것 같습니다.
다시 생각해보니 테스트용 DB의 `DataSource`를 포함한 `JdbcTemplate`을 주입하면 프로덕션 영향 없이 테스트할 수 있겠네요.
이 아이디어를 이제서야 떠올렸어요 🥺

근데, `Repository`가 실제 DB를 사용하기에 `Controller` 테스트가 난감했고, 따라서 DB가 아닌 컬렉션을 사용해서 영속성을 흉내내는 `FakeRepository`를 사용해서 `Controller`를 테스트한건데, 사실 처음부터 `Repository`의 `JdbcTemplate`을 테스트용 DB로 주입하고, 그것을 또 `Controller`에 주입하면 쉽게 `Controller`를 테스트 할 수 있었을 것 같네요.

`FakeRepository`를 사용하면 DB에 의존하지 않고, 연결 및 초기화 비용 없이 빠르게 테스트 할 수 있는것이 장점 같은데, 사실 명확함을 위해서라면 방금 떠오른 방법처럼 테스트용 DB를 주입해서 테스트하는 것이 더 좋을 것 같다는 생각이 들어요 :)

sinb57 commented on src/main/java/roomescape/repository/impl/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,66 @@
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(query, new String[]{"id"});
+            ps.setString(1, reservation.getName());
+            ps.setObject(2, reservation.getDate());
+            ps.setLong(3, reservation.getTimeId());
```
매번 순서와 인덱스를 신경써야 할텐데요, JdbcTemplate의 다른 구현체를 활용해서 해결해보면 어떨까요?

Hacanna42 replied:
`SimpleJdbcInsert` 에 대해 말씀해주시는건가요? 사실 `JdbcTemplate`과 `KeyHolder`을 사용하면서 가독성이 정말 안좋다고 생각했어요. 그래서 이 구현체를 활용해보려고 하다가, 솔직하게 말하자면 요구사항을 이미 완료했기에 굳이 구현체를 변경하지 않은 것 같습니다.

하지만 지금은 학습 목적이기 때문에, `SimpleJdbcInsert`로 해결해보겠습니다 :)
놓친 부분 짚어주셔서 감사합니다!

sinb57 replied:
`SimpleJdbcInsert`도 좋지만 `NamedParameterJdbcTemplate`도 찾아보길 바랬어요~ㅎ

Hacanna42 replied:
`NamedParameterJdbcTemplate` 는 처음 들어보네요! 한번 찾아보고 학습해볼게요 :)
좋은 키워드 제시해주셔서 감사합니다 ~!

sinb57 commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,43 @@
+    @PostMapping
+    ResponseEntity<ReservationTime> create(@RequestBody ReservationTime reservationTime) {
+        ReservationTime createdReservationTime = reservationTimeRepository.createReservationTime(reservationTime);
+        return ResponseEntity.ok(createdReservationTime);
```
ResponseEntity를 사용한 이유가 궁금해요~

Hacanna42 replied:
`ResponseEntity`는 HTTP 응답 전체를 직접 제어할 수 있도록 해주는 Spring의 추상화 도구로, 상태 코드, 헤더, 바디를 명시적으로 설정할 수 있고, 제일 중요한 것은 성숙하고 널리 사용되는 관습적인 방식이라고 판단했어요.

저는 이런 점에서 `ResponseEntity`가 이미 잘 만들어진 표준이라고 판단했고, 이를 사용하지 않는 것은 마치 바퀴를 다시 발명하는 일처럼 느껴졌습니다.

sinb57 replied:
`ResponseEntity.ok(##)`와 같이 메서드를 내부를 들여다보아야 상태코드를 확인할 수 있는 반면, `@ResponseStatus` 어노테이션을 통해 상태코드를 커스텀하면 메서드 시그니처에서 곧바로 상태코드를 확인할 수 있어요. 이에 대해 히스타는 어떻게 생각하시는지 궁금해요~

Hacanna42 replied:
`@ResponseStatus` 라는 어노테이션을 이프 덕분에 처음 알았네요. 감사합니다!
`@ResponseStatus`를 사용하면 선언적으로 상태 코드를 정의할 수 있네요.

위 코드를 `ResponseStatus`를 사용하도록 개선하면 
```java
@ResponseStatus(HttpStatus.OK)
@PostMapping
ReservationTime create(@RequestBody ReservationTime reservationTime) {
    return reservationTimeRepository.createReservationTime(reservationTime);
}
```

이런식으로 할 수 있을 것 같은데, 메서드가 상당히 간결해지고 상태 코드도 바로 보여서 직관적이네요.
`Body`만 정의하면 되는 응답이라면 이런 방식으로 안 할 이유가 없어 보여요! 아마 자주 애용하게 될 것 같습니다 :)

하지만 `Body`가 아니라 `Header` 등의 추가적인 정보도 정의해야 하는 일이 있다면, HTTP 응답을 제어할 수 있는 도구인 `ResponseEntity`를 사용할 것 같아요. 

추가적으로, 하나의 컨트롤러에서 어떤 메서드는 추가 정의가 필요해서 `ResponseEntity`를 사용하고, 어떤 메서드는 `Body`만 있으면 돼서 `@ResponseStatus` 를 사용했다면 오히려 일관성이 떨어져서 모두 `ResponseEntity`로 통일하게 될 것 같습니다~

어떤 건 `ResponseEntity.ok` 고 어떤 건 그냥 `return + @ResponseStatus(HttpStatus.OK)` 면 오히려 헷갈리게 될 것 같네요.

sinb57 replied:
히스타 의견에 동의해요~ 필요에 따라선 `ResponseEntity`를 사용해야 할 때가 분명 존재하고, 그런 경우 여러 방식이 혼재되면 혼란이 야기되기도 해요. 다만 헤더를 커스텀하는 등의 행위가 잦지 않아서, 아직은 `ResponseEntity`를 그리 잘 사용하고 있지 않네요ㅎ

sinb57 commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,43 @@
+import roomescape.domain.ReservationTime;
+import roomescape.repository.ReservationTimeRepository;
+
+@RestController
```
`@RestController`와 `@Controller`의 차이가 무엇인가요?

Hacanna42 replied:
> `@Controller`는 무엇인가요?

`@Controller` 는 `@Component` 어노테이션의 특수화로, 해당 클래스를 `Controller`로 사용한다고 Spring에게 알려줍니다.

그래서 얻을 수 있는 것들은:
1. Spring에 Bean으로 등록됨.
2. 요청 처리 대상이 됨. (`@RequestMapping` - 이것이 `DispatcherServlet`과도 밀접한 연관이 있는 것으로 알고 있어요. 학습 예정입니다.) 
3. 반환 값이 문자열인 경우, 그것을 문자열이 아닌 `View`로 인식합니다. (내부적으로 `ViewResolver`가 작동해서, 이름에 맞는 `View`를 불러오는 것으로 알고 있습니다.)

<br>

> `@RestController`는 무엇인가요?

`@RestController` 는 `@Controller`의 특수화로, 정확하게 <`@Controller` + `@ResponseBody`> 의 형태입니다.
`@ResponseBody` 는 모든 반환 값을 응답 본문(body)에 직렬화해서 전송합니다. 즉 반환값이 `ViewResolver`에 의해 `View`로 변환되는 형태가 아닙니다.

그러면 왜 굳이 `@Controller`와 `@ResponseBody` 를 두개 다 쓰지 않고 `@RestController` 라고 쓸까요?
(정확한 정보는 아니지만) Spring docs에서 얼핏 본 기억이 나지만, 그냥 `@RestController` 을 사용함으로써 REST API 관련 컨트롤러임을 직관적으로 전할 수 있고, 추후 Spring이 해당 어노테이션에 업데이트를 적용할 때 즉각적으로 적용될 수 있다고 합니다.

sinb57 replied:
> 그러면 왜 굳이 @Controller와 @ResponseBody 를 두개 다 쓰지 않고 @RestController 라고 쓸까요?

Spring 오픈소스를 살펴보면, 보일러 플레이트를 제거하기 위해 도입되었음을 알 수 있어요~
https://github.com/spring-projects/spring-framework/issues/15440

Hacanna42 replied:
오호 더 구체적인 이유가 있었군요. 감사합니다 :)

sinb57 commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,43 @@
+public class ReservationTimeController {
+    private final ReservationTimeRepository reservationTimeRepository;
+
+    @Autowired
+    public ReservationTimeController(ReservationTimeRepository reservationTimeRepository) {
```
`@Autowired`는 어떠한 이유로 생성자에 붙이셨나요? `@Autowired`는 주로 어떨때 사용할까요?

Hacanna42 replied:
> `@Autowired`는 어떠한 이유로 생성자에 붙이셨나요?

처음엔 필드에 붙였었는데, 생성자에 붙이면 생성 시점에 의존 관계가 결정되기 때문에 불변성이 보장되고, 필요에 따라 수동적으로 의존성을 주입할 수 있기에 생성자에 붙이는 것이 좀 더 바람직하다고 판단했습니다.

> `@Autowired`는 주로 어떨때 사용할까요?

Spring의 의존 관계를 직접 설정하도록 해서, 개발자가 의존 관계에 얽매이지 않고 구현에만 집중할 수 있게 합니다.
즉 `@Autowired`를 사용하면 의존성 주입에 대한 걱정을 덜 할 수 있기에, 수동 의존성 주입이 필요하지 않으면 거의 사용하는 것 같습니다.
이것이 기존 개발자가 의존 관계를 설정했던 것에 반하여, 객체의 제어권이 Spring Framework에게 넘어갔다고 하여 제어의 역전(IoC) 라고 합니다.

sinb57 replied:
`@Autowired`를 붙일 필요가 없음을 전하고 싶었는데, 이미 인지하고서 의도적으로 붙인 것이군요!
본인의 기준을 가진 모습이 좋네요ㅎ

sinb57 commented on src/test/java/roomescape/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+        String givenName = "히스타";
+        Long givenTimeId = 1L;
+
+        create(givenName, givenDate, givenTimeId, givenTime);
```
중복 제거를 위함이어도, 테스트 간의 결합은 지양하는 것이 좋아요~

Hacanna42 replied:
동의합니다. 학습 테스트용으로 제공한 코드에서 저렇게 중복 코드를 제거하길래 따라해봤는데...
안 그래도 `create` 테스트 메서드가 한번 변경되었었는데, 다른 테스트 메서드도 수정되는 것이 안좋더군요.

해당 부분 수정하겠습니다.

sinb57 commented on src/test/java/roomescape/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+    @Test
+    void delete() {
+        // given
+        create("히스타", LocalDate.now(), 1L, LocalTime.now());
```
```suggestion
        create("히스타", LocalDate.now(), 2L, LocalTime.now());
```
왜 timeId를 바꾸면 테스트가 실패할까요~?

Hacanna42 replied:
~~`create()` 메서드가 항상 `1L`의 `id`로 `ReservationTime`을 생성하기 때문입니다.
테스트 간의 결합이 안좋은 이유를 바로 설명해주셨네요. 설득되었습니다! 감사합니다 :)~~

- 수정 (25-04-25)
결국 `FakeRepository`가 `AtomicLong`을 사용하여 id를 1부터 생성하는데, 테스트 코드에서 ID를 지정할 수 있다고 착각해버렸네요. 

오해 소지가 없도록 ID를 직접 지정할 수 없도록 수정하겠습니다 :)

sinb57 replied:
테스트라는 환경상 임의의 ID로 저장하는 경우가 분명 존재할 텐데요. ID를 직접 지정할 수 없도록 하는 것이 과연 옳은지 궁금해요~

Hacanna42 replied:
안 그래도 해당 부분에 대해, **"테스트 코드니까 ID(pk)를 직접 지정할 수 있도록 하면 더 유연하지 않을까?"** 라는 고민을 했었는데요.

1. 테스트 환경이지만 `FakeRepository` 객체는 실제 프로덕션 `Repository`(DB를 이용하는)와 가능한 한 유사해야 한다고 생각했습니다.
2. 그런데 프로덕션에서 임의의 ID(pk)로 저장하는 경우는 일어나지 않을 일이라고 판단했습니다. (Auto increment가 적용된 pk니까)
3. 그러면 테스트 코드에서 임의의 ID로 저장해서 테스트 해야하는 경우가 존재할까? **(= 아니요)**

따라서 ID를 직접 지정할 수 없도록 했습니다. ID를 직접 지정해야 하는 테스트를 작성해야 할 일이 있을까요?

sinb57 replied:
테스트 환경은 운영 환경의 특정 시점을 재현할 수 있어야 해요. ID를 직접 지정해야 하는 경우가, 지금 상상해보기론 다음과 같은 상황이 떠오르네요. 특정 N배수의 순서로 가입한 사용자에게 무언가 혜택을 주고자 해요. 그렇다면 가입시 부여되는 ID를 활용하면 될 것이고, ID 값이 N배수인지 확인하여 로직을 수행하면 될 거예요. 이를 테스트하고자 한다면, N배수의 ID인지 아닌지 두가지 경우의 수만 테스트하면 될텐데요. 그에 반해 히스타의 방식은 N배수의 ID를 지닌 엔티티를 만들기 위해 N배수 번째까지의 엔티티를 만들어야 해요. 이처럼 ID가 로직에 영향을 미치는 경우는, ID를 직접 지정하여 테스트를 작성하곤 해요.

지금의 요구조건으로는 히스타의 방식에 문제는 없으나, ID를 직접 지정할 수 없어야 한다는 관념을 없앴으면 하는 바램에서 질문 드려보았어요~

Hacanna42 replied:
그렇네요. 재현 관점에서 테스트 코드에서 ID를 지정할 수 있도록 하면 좋겠네요. 구체적인 상황 제시해주셔서 이해가 됐어요!

감사합니다 !!

sinb57 reviewed (changes requested):
안녕하세요 히스타~ 미션 진행하느라 고생 많으셨어요!
간단히 코멘트 남겨두었으니 확인 부탁드려요~!!

Hacanna42 committed:
refactor: Controller가 요청을 DTO로 받도록 수정

Hacanna42 committed:
refactor: Optional로 에러 처리 및 Service 계층 분리

- 이미 저장된 예약의 시간 값이 관리자에 의해 삭제된 경우의 예외 처리
- 예약 자체의 시간 ID가 잘못된 경우의 예외 처리
- Service 계층 분리

Hacanna42 committed:
refactor: SimpleJdbcInsert 적용으로 가독성 개선

Hacanna42 committed:
test: 테스트간의 결합 삭제

sinb57 reviewed (changes requested):
안녕하세요 히스타~ 미션 진행하느라 고생많으셔요
코멘트 남겨놓았으니 확인 부탁드려요~ㅎ

Hacanna42 committed:
feat: DB가 휘발되지 않도록 Embedded 모드로 변경

- H2의 모드를 Embedded로 설정해서, 프로그램 재시작시 초기화되지 않도록 설정.
- 테스트 코드에서 프로덕션의 Embedded DB를 사용하지 않도록 properties 설정.

sinb57 commented on src/test/java/roomescape/domain/ReservationTimeTest.java:
```diff
@@ -0,0 +1,18 @@
+
+class ReservationTimeTest {
+    @Test
+    void validate() {
+        // given
+        Supplier<ReservationTime> supplier = () -> new ReservationTime(1L, null);
+
+        // when & then
+        Assertions.assertThatNullPointerException().isThrownBy(
+                supplier::get
+        );
+    }
```
Supplier 사용은 `given / when & then` 구분 짓기 위함인가요?? 아래와 같이 합치지 않은 이유가 궁금해요~

```suggestion
    void validate() {
        Assertions.assertThatNullPointerException().isThrownBy(
                new ReservationTime(1L, null);
        );
    }
```

Hacanna42 replied:
네. 생성 자체가 테스트 로직일 경우 `given & when & then` 이 모두 합쳐져 있는 게 복잡해보여서
`given`과 `when & then`을 분리해봤어요.

사실 비교해서 보니까... 기존 방식에서 더 간결해졌다는 느낌이 안 들긴 하네요..

sinb57 replied:
히스타는 `given & when & then`의 각 단계가 무엇을 요구한다고 생각하시나요? 지금의 `given`은 어떠한 의미를 담고 있나요? 저는 `given`이 테스트에서 꽤나 중요한 요소라고 생각하는데요. `given`에 따라 테스트가 성공할 수도 실패할 수 있어요. 지금의 `validate()` 테스트에서는 검증하고자 하는 요소인 startAt 파라미터의 null 여부겠지요. 따라서 저라면 아래와 같이 테스트를 작성했을 듯 해요. 그치만 이런 방식은 취향에 따라 갈리기 때문에, 히스타가 편한 방식으로 해주시면 돼요~ㅎ

```suggestion
    void validate() {
        // given
        LocalTime startAt = null;

        // when & then
        assertThatNullPointerException()
                .isThrownBy(() -> new ReservationTime(1L, startAt));
    }```

sinb57 reviewed (changes requested):
안녕하세요 히스타~ 주말에도 미션 진행하느라 고생 많으셨어요!
코멘트 확인해주시고 답변 남겨주세요~

sinb57 reviewed (changes requested):
안녕하세요 히스타~ 주말동안 미션 진행하느라 고생 많으셨어요!
현재 미션에서 요구하는 학습 수준은 충분히 만족하셔서 이만 머지하려고 해요~
코멘트 남겨놓았으니 확인해주시고요. 별다른 conversation이 없다면 다음 요청 때 머지하겠습니다!

sinb57 reviewed (approved):
안녕하세요 히스타~ 방탈출 예약 관리 미션 진행하느라 고생 많으셨어요!
레벨2의 첫번째 미션을 함께하여 영광이었습니다!
앞으로의 우테코 생활을 응원할게요~ 다음 미션도 홧팅 하세요~!!
